### PR TITLE
Passwordless

### PR DESCRIPTION
**Pull request by siddharthmudgal**
## Summary of change
Passwordless modules enabled for the mysql plugin

## Related issues
- [For zenhub](https://github.com/supertokens/for-zenhub/issues/62)
- supertokens/supertokens-core#344

## Test Plan
Testing for this code is covered via the core-passwordless tests.

## Documentation changes
No documentation required since the changes are internal

## Checklist for important updates
- [x] Changelog has been updated
- [x] `pluginInterfaceSupported.json` file has been updated (if needed)
- [x] Changes to the version if needed
   - In `build.gradle`
- [x] Had installed and ran the pre-commit hook
- [x] If there are new dependencies that have been added in `build.gradle`, please make sure to add them in `implementationDependencies.json`.
- [x] Issue this PR against the latest non released version branch.
   - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
   - If no such branch exists, then create one from the latest released branch.

### CODE REVIEW

File: src/main/java/io/supertokens/storage/mysql/config/MySQLConfig.java [hunk 1]
```diff
@@ -68,6 +68,15 @@
...
 
+    @JsonProperty
+    private String passwordless_users_table = null;
+
+    @JsonProperty
+    private String passwordless_device_table = null;
+
+    @JsonProperty
+    private String passwordless_codes_table = null;
+
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/mysql/config/MySQLConfig.java:79]: These configs are not needed. They were there for older tables, but we stopped adding them for newer ones since someone can use the prefix config to change table names anyway

File: src/main/java/io/supertokens/storage/mysql/config/MySQLConfig.java [hunk 2]
```diff
@@ -250,6 +259,34 @@ public String getThirdPartyUsersTable() {
...
+        return addPrefixToTableName(tableName);
+
+    }
+
+    public String getPasswordlessDevicesTable() {
+        String tableName = "passwordless_devices";
+
+        if (passwordless_device_table != null)
+            return passwordless_device_table;
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/mysql/config/MySQLConfig.java:276]: No need to do this. Also, I dislike `if` statements without a bracket. Please be sure to read the contributing guide we have!

File: src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java [hunk 3]
```diff
@@ -0,0 +1,570 @@
+/*
+ *    Copyright (c) 2020, VRAI Labs and/or its affiliates. All rights reserved.
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java:2]: 2021

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

**Comment by rishabhpoddar** [comment]
@porcellus i have reviewed just a few parts of the PR. Please do a more thorough review for all the functions implemented.

File: src/main/java/io/supertokens/storage/mysql/Start.java [hunk 4]
```diff
@@ -898,4 +902,305 @@ public void setJWTSigningKey_Transaction(TransactionConnection con, JWTSigningKe
...
+    @Override
+    public void createUser(io.supertokens.pluginInterface.passwordless.UserInfo user) throws StorageQueryException,
+            DuplicateEmailException, DuplicatePhoneNumberException, DuplicateUserIdException {
+        try {
+            PasswordlessQueries.createUser(this, user);
+        } catch (StorageTransactionLogicException e) {
+            // TODO: review these exceptions
+            String message = e.actualException.getMessage();
+            if (message.matches(
+                    ".*Duplicate.*entry.*" + Config.getConfig(this).getPasswordlessUsersTable() + ".*PRIMARY\\'") ||
```
Comment (porcellus) [inline src/main/java/io/supertokens/storage/mysql/Start.java:1155]: I like the change to regex, but I think they need a bit of cleanup: they allow different things and there are some strange things (escaped "_")

Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/mysql/Start.java:1155]: Not sure why are are using "*". Are you sure it won't give false positives? If yes, we should consider making other checks with "*" as well..?

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java [hunk 5]
```diff
@@ -0,0 +1,570 @@
...
+            }
+            return null;
+        });
+    }
+
+    public static PasswordlessDevice getDevice_Transaction(Start start, Connection con, String deviceIdHash)
+            throws StorageQueryException, SQLException {
+
+        String QUERY = "SELECT device_id_hash, email, phone_number, failed_attempts FROM "
+                + Config.getConfig(start).getPasswordlessDevicesTable() + " WHERE device_id_hash = ?";
```
Comment (porcellus) [inline src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java:103]: We need to lock the device using FOR UPDATE

File: src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java [hunk 6]
```diff
@@ -0,0 +1,570 @@
...
+
+        List<String> deviceIdHashes = new ArrayList<>();
+        try (PreparedStatement pst = con.prepareStatement(QUERY)) {
+            pst.setString(1, email);
+            ResultSet result = pst.executeQuery();
+            while (result.next()) {
+                deviceIdHashes.add(result.getString("device_id_hash"));
+            }
+        }
+        deleteRowsByDeviceIdHashList_Transaction(con, Config.getConfig(start).getPasswordlessDevicesTable(),
```
Comment (porcellus) [inline src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java:167]: It can be just a single query.

File: src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java [hunk 7]
```diff
@@ -0,0 +1,570 @@
...
+        }
+    }
+
+    public static void createCode(Start start, PasswordlessCode code)
+            throws StorageTransactionLogicException, StorageQueryException {
+        start.startTransaction(con -> {
+            Connection sqlCon = (Connection) con.getConnection();
+
+            try {
+                if (PasswordlessQueries.getDevice_Transaction(start, sqlCon, code.deviceIdHash) == null) {
```
Comment (porcellus) [inline src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java:189]: This should be done by the foreign key constraint.

File: src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java [hunk 8]
```diff
@@ -0,0 +1,570 @@
...
+            ResultSet result = pst.executeQuery();
+
+            if (result.next()) {
+                return UserInfoRowMapper.getInstance().mapOrThrow(result);
+            }
+        }
+        return null;
+    }
+
+    private static void deleteRowsByDeviceIdHashList_Transaction(Connection con, String tableName,
```
Comment (porcellus) [inline src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java:496]: I don't think we need this.

**Review by porcellus — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/storage/mysql/config/MySQLConfig.java (hunk 1 above)
Comment (siddharthmudgal) [reply src/main/java/io/supertokens/storage/mysql/config/MySQLConfig.java]: roger that

File: src/main/java/io/supertokens/storage/mysql/config/MySQLConfig.java (hunk 2 above)
Comment (siddharthmudgal) [reply src/main/java/io/supertokens/storage/mysql/config/MySQLConfig.java]: noted.

File: src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java (hunk 3 above)
Comment (siddharthmudgal) [reply src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java]: done

File: src/main/java/io/supertokens/storage/mysql/Start.java (hunk 4 above)
Comment (siddharthmudgal) [reply src/main/java/io/supertokens/storage/mysql/Start.java:1155]: I am happy to discuss more about regular expressions, I really enjoy building them! It shouldn't give false positives, since it needs to contain "Duplicate entry passwordless_users PRIMARY" keywords in these sequence. But again, we can discuss and improve it.

Comment (siddharthmudgal) [reply src/main/java/io/supertokens/storage/mysql/Start.java:1155]: Thank you, I love regular expressions. Sure, happy to discuss any changes. Generally symbols are escaped in regular expressions, to avoid any confusion this regex contains escaped underscores.

Comment (porcellus) [reply src/main/java/io/supertokens/storage/mysql/Start.java:1155]: Hmm, to be honest, seeing the escape sequence was more confusing than seeing just the underscore as that has no special meaning, so I was unsure if I forgot about something or if this was in some weird extension. :)

Comment (porcellus) [reply src/main/java/io/supertokens/storage/mysql/Start.java:1155]: The sequence is guaranteed, but it still matches strings that the exception detection part of other methods won't. So while I think doing this in a single regex is good, we should keep the pattern the same.

Comment (siddharthmudgal) [reply src/main/java/io/supertokens/storage/mysql/Start.java:1155]: so from what I understand, we are not making any changes in this section right ?

File: src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java (hunk 8 above)
Comment (siddharthmudgal) [reply src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java]: roger.

File: src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java (hunk 6 above)
Comment (siddharthmudgal) [reply src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java]: done.

File: src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java (hunk 5 above)
Comment (siddharthmudgal) [reply src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java]: @rishabhpoddar @porcellus added : "LOCK IN SHARE MODE" instead of "FOR UPDATE". Does that sound good ?

File: src/main/java/io/supertokens/storage/mysql/Start.java (hunk 4 above)
Comment (porcellus) [reply src/main/java/io/supertokens/storage/mysql/Start.java:1155]: No, I meant we need to change it to match what we have originally or copy the old way of exception detection. Either one is fine with me; copying is better for consistency.

File: src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java (hunk 5 above)
Comment (porcellus) [reply src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java]: No, we want FOR UPDATE here.

File: src/main/java/io/supertokens/storage/mysql/Start.java (hunk 4 above)
Comment (siddharthmudgal) [reply src/main/java/io/supertokens/storage/mysql/Start.java:1155]: understood, thanks!

File: src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java (hunk 7 above)
Comment (siddharthmudgal) [reply src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java:177]: noted.

File: src/main/java/io/supertokens/storage/mysql/Start.java [hunk 9]
```diff
@@ -310,14 +314,14 @@ public void createNewSession(String sessionHandle, String userId, String refresh
...
-        }
-    }
+//    @Override
+//    public void deleteSessionsOfUser(String userId) throws StorageQueryException {
+//        try {
+//            SessionQueries.deleteSessionsOfUser(this, userId);
+//        } catch (SQLException e) {
+//            throw new StorageQueryException(e);
+//        }
+//    }
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/mysql/Start.java:324]: Why is this commented out? Either merge the latest release into this branch, or then remove the comment.

File: src/main/java/io/supertokens/storage/mysql/Start.java [hunk 10]
```diff
@@ -934,4 +911,315 @@ public void setJWTSigningKey_Transaction(TransactionConnection con, JWTSigningKe
...
+    public void createDeviceWithCode(@Nullable String email, @Nullable String phoneNumber, @NotNull String linkCodeSalt,
+            PasswordlessCode code) throws StorageQueryException, DuplicateDeviceIdHashException,
+            DuplicateCodeIdException, DuplicateLinkCodeHashException {
+        if (email == null && phoneNumber == null) {
+            throw new IllegalArgumentException("Both email and phoneNumber can't be null");
+        }
+        try {
+            PasswordlessQueries.createDeviceWithCode(this, email, phoneNumber, linkCodeSalt, code);
+        } catch (StorageTransactionLogicException e) {
+            // TODO: review these error messages
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/mysql/Start.java:1022]: If this is done, then remove this TODO. Also, I thought you were going to use regex matching - why was that removed?

File: src/main/java/io/supertokens/storage/mysql/Start.java [hunk 11]
```diff
@@ -934,4 +911,315 @@ public void setJWTSigningKey_Transaction(TransactionConnection con, JWTSigningKe
...
+            PasswordlessQueries.createCode(this, code);
+        } catch (StorageTransactionLogicException e) {
+            // TODO: review these exceptions
+            String message = e.actualException.getMessage();
+
+            if (e.actualException instanceof UnknownDeviceIdHash) {
+                throw (UnknownDeviceIdHash) e.actualException;
+            }
+            if (message.contains(
+                    "Cannot add or update a child row: a foreign key constraint fails (`supertokens`.`passwordless_codes`, CONSTRAINT `passwordless_codes_ibfk_1` FOREIGN KEY (`device_id_hash`) REFERENCES `passwordless_devices` (`device_id_hash`) ON DELETE CASCADE ON UPDATE CASCA)")) {
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/mysql/Start.java:1122]: This message will probably fail if the table is renamed via a prefix? See how other foreign key constrain violations are checked. Like [here](https://github.com/supertokens/supertokens-mysql-plugin/blob/e0e17c9eb2b3081d8acdbc70b279aef4764c1978/src/main/java/io/supertokens/storage/mysql/Start.java#L522)

File: src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java [hunk 12]
```diff
@@ -0,0 +1,539 @@
...
+            try {
+                PasswordlessQueries.createCode_Transaction(start, sqlCon, code);
+                sqlCon.commit();
+            } catch (SQLException e) {
+
+                String message = e.getMessage();
+                if (message.contains(
+                        "Cannot add or update a child row: a foreign key constraint fails (`supertokens`.`passwordless_codes`, CONSTRAINT `passwordless_codes_ibfk_1` FOREIGN KEY (`device_id_hash`) REFERENCES `passwordless_devices` (`device_id_hash`) ON DELETE CASCADE ON UPDATE CASCA)")) {
+                    throw new StorageTransactionLogicException(new UnknownDeviceIdHash());
+                }
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java:187]: You are already checking this in Start.java.. no need to check it here.

File: src/main/java/io/supertokens/storage/mysql/Start.java [hunk 13]
```diff
@@ -934,4 +911,315 @@ public void setJWTSigningKey_Transaction(TransactionConnection con, JWTSigningKe
...
+    }
+
+    @Override
+    public void updateUserEmail_Transaction(TransactionConnection con, String userId, String email)
+            throws StorageQueryException, UnknownUserIdException, DuplicateEmailException {
+        Connection sqlCon = (Connection) con.getConnection();
+        try {
+            PasswordlessQueries.updateUserEmail_Transaction(this, sqlCon, userId, email);
+        } catch (SQLException e) {
+            // TODO: review these exceptions
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/mysql/Start.java:1197]: If this is reviewed, please remove TODO

File: src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java [hunk 14]
```diff
@@ -0,0 +1,539 @@
...
+            throws SQLException, UnknownUserIdException {
+        String QUERY = "UPDATE " + Config.getConfig(start).getPasswordlessUsersTable()
+                + " SET email = ? WHERE user_id = ?";
+
+        try (PreparedStatement pst = con.prepareStatement(QUERY)) {
+            pst.setString(1, email);
+            pst.setString(2, userId);
+            if (pst.executeUpdate() != 1) {
+                throw new UnknownUserIdException();
+            }
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java:285]: You should make this function return the number of rows updated. And in start.java, we check if it's not 1, then we throw `UnknownUserIdException`. This is cause this file is only responsible for running the queries and not mapping the query result to the exceptions thrown from the interface functions.

File: src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java [hunk 15]
```diff
@@ -0,0 +1,539 @@
...
+            throws SQLException, UnknownUserIdException {
+        String QUERY = "UPDATE " + Config.getConfig(start).getPasswordlessUsersTable()
+                + " SET phone_number = ? WHERE user_id = ?";
+
+        try (PreparedStatement pst = con.prepareStatement(QUERY)) {
+            pst.setString(1, phoneNumber);
+            pst.setString(2, userId);
+            if (pst.executeUpdate() != 1) {
+                throw new UnknownUserIdException();
+            }
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java:299]: You should make this function return the number of rows updated. And in start.java, we check if it's not 1, then we throw `UnknownUserIdException`. This is cause this file is only responsible for running the queries and not mapping the query result to the exceptions thrown from the interface functions.

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/storage/mysql/Start.java (hunk 10 above)
Comment (porcellus) [reply src/main/java/io/supertokens/storage/mysql/Start.java:1022]: We've discussed that it may be better for consistency if the checked worked the same way as it does right now and to discuss and maybe update the way we check exceptions separately.

File: src/main/java/io/supertokens/storage/mysql/Start.java (hunk 9 above)
Comment (porcellus) [reply src/main/java/io/supertokens/storage/mysql/Start.java:324]: + Revert the deleted functions. I think the issue is that the plugin interface repo doesn't have the newest things merged.

File: src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java (hunk 14 above)
Comment (porcellus) [reply src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java:285]: I think this mistake is also present in the in-memory implementation, we should update it there as well.

**Review by porcellus — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/storage/mysql/Start.java (hunk 9 above)
Comment (siddharthmudgal) [reply src/main/java/io/supertokens/storage/mysql/Start.java:324]: I am on the latest passwordless branch of interface, after the recent updates I had to remove some interface implementations. I removed the comment

File: src/main/java/io/supertokens/storage/mysql/Start.java (hunk 11 above)
Comment (siddharthmudgal) [reply src/main/java/io/supertokens/storage/mysql/Start.java:1122]: noted.

File: src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java (hunk 14 above)
Comment (rishabhpoddar) [reply src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java:285]: @porcellus please do.

File: src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java (hunk 12 above)
Comment (siddharthmudgal) [reply src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java:187]: Actually it makes more sense to add the check here instead of start.java, because in-memory has been done that way. Making this change for consistency.

if ( message.contains("foreign key constraint fails") &&
                        message.contains("passwordless_codes") &&
                        message.contains("device_id_hash") )
                {
                    throw new StorageTransactionLogicException(new UnknownDeviceIdHash());
                }

File: src/main/java/io/supertokens/storage/mysql/Start.java (hunk 13 above)
Comment (siddharthmudgal) [reply src/main/java/io/supertokens/storage/mysql/Start.java:1197]: done.

File: src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java (hunk 14 above)
Comment (siddharthmudgal) [reply src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java:285]: Shouldn't this be targeted in a different tracker for in-memory. I have updated it for sql

File: src/main/java/io/supertokens/storage/mysql/Start.java (hunk 10 above)
Comment (rishabhpoddar) [reply src/main/java/io/supertokens/storage/mysql/Start.java]: okay!

File: src/main/java/io/supertokens/storage/mysql/Start.java [hunk 16]
```diff
@@ -1112,16 +1138,11 @@ public void createCode(PasswordlessCode code) throws StorageQueryException, Unkn
         try {
             PasswordlessQueries.createCode(this, code);
         } catch (StorageTransactionLogicException e) {
-            // TODO: review these exceptions
             String message = e.actualException.getMessage();
 
             if (e.actualException instanceof UnknownDeviceIdHash) {
                 throw (UnknownDeviceIdHash) e.actualException;
             }
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/mysql/Start.java:1145]: PasswordlessQueries should not throw `UnknownDeviceIdHash`. The message check should still happen in this file

File: src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java [hunk 17]
```diff
@@ -181,8 +181,8 @@ public static void createCode(Start start, PasswordlessCode code)
             } catch (SQLException e) {
 
                 String message = e.getMessage();
-                if (message.contains(
-                        "Cannot add or update a child row: a foreign key constraint fails (`supertokens`.`passwordless_codes`, CONSTRAINT `passwordless_codes_ibfk_1` FOREIGN KEY (`device_id_hash`) REFERENCES `passwordless_devices` (`device_id_hash`) ON DELETE CASCADE ON UPDATE CASCA)")) {
+                if (message.contains("foreign key constraint fails") && message.contains("passwordless_codes")
+                        && message.contains("device_id_hash")) {
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java:185]: - The message check should happen in Start.java and not here
- Please see how we check for foreign key violations in other places. We don't actually check for table names. And if we must do that here, we should check the full table name using `Config.getConfig(start).getPasswordlessCodesTable()`

File: src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java (hunk 12 above)
Comment (rishabhpoddar) [reply src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java]: Okay.

File: src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java (hunk 14 above)
Comment (rishabhpoddar) [reply src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java]: Yes. I have added a comment in the core PR as well for this: https://github.com/supertokens/supertokens-core/pull/344#discussion_r773258140

File: src/main/java/io/supertokens/storage/mysql/Start.java [hunk 18]
```diff
@@ -1192,16 +1212,23 @@ public void updateUserEmail_Transaction(TransactionConnection con, String userId
...
-            // TODO: review these exceptions
+            int updated_rows = PasswordlessQueries.updateUserEmail_Transaction(this, sqlCon, userId, email);
+            if (updated_rows != 1) {
+                throw new UnknownUserIdException();
+            }
+        } catch (UnknownUserIdException | SQLException e) {
+
+            if (e instanceof UnknownUserIdException) {
+                throw new UnknownUserIdException();
+            }
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/mysql/Start.java:1223]: You don't need to catch `UnknownUserIdException` and then throw the exact same thing again. Just don't catch `UnknownUserIdException` :)

File: src/main/java/io/supertokens/storage/mysql/Start.java [hunk 19]
```diff
@@ -1210,9 +1237,18 @@ public void updateUserPhoneNumber_Transaction(TransactionConnection con, String
...
+
+            if (updated_rows != 1) {
+                throw new UnknownUserIdException();
+            }
+
+        } catch (UnknownUserIdException | SQLException e) {
+
+            if (e instanceof UnknownUserIdException) {
+                throw new UnknownUserIdException();
+            }
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/mysql/Start.java:1250]: You don't need to catch `UnknownUserIdException` and then throw the exact same thing again. Just don't catch `UnknownUserIdException` :)

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 20]
```diff
@@ -13,6 +13,7 @@ to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 - Delete user functionality
 
+
```
Comment (rishabhpoddar) [inline CHANGELOG.md:16]: You also need to modify the `pluginInterfaceSupported.json` file to replace "2.10" with "2.11"

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 20 above)
Comment (rishabhpoddar) [reply CHANGELOG.md:16]: And need to modify the build.gradle to have version 1.12

File: src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java (hunk 17 above)
Comment (siddharthmudgal) [reply src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java:185]: noted.

File: src/main/java/io/supertokens/storage/mysql/Start.java (hunk 18 above)
Comment (siddharthmudgal) [reply src/main/java/io/supertokens/storage/mysql/Start.java:1223]: Agreed, I missed that.

File: src/main/java/io/supertokens/storage/mysql/Start.java (hunk 19 above)
Comment (siddharthmudgal) [reply src/main/java/io/supertokens/storage/mysql/Start.java:1250]: Agreed, I missed that.

File: CHANGELOG.md (hunk 20 above)
Comment (siddharthmudgal) [reply CHANGELOG.md:16]: both done.

File: src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java [hunk 21]
```diff
@@ -181,7 +181,8 @@ public static void createCode(Start start, PasswordlessCode code)
             } catch (SQLException e) {
 
                 String message = e.getMessage();
-                if (message.contains("foreign key constraint fails") && message.contains("passwordless_codes")
+                if (message.contains("foreign key constraint fails")
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java:184]: If you see other places, we do `e.getMessage().contains("foreign key") && ...`. So why not do that here too? Just to be consistent.

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java (hunk 21 above)
Comment (siddharthmudgal) [reply src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java:184]: I got this from : "String message = e.actualException.getMessage();" and I felt this is a much cleaner way to do it. Let me know what needs to be done.

Comment (siddharthmudgal) [reply src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java:184]: Proceeding with updating just the ones with message = e.getMessage(); Will leave the e.actualException.getMessage() to be updated in another PR.

Comment (rishabhpoddar) [reply src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java]: I should have been clearer.. My issue is with using "foreign key constraint fails" as the match criteria instead of just "foreign key"

File: src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java [hunk 22]
```diff
@@ -180,10 +180,9 @@ public static void createCode(Start start, PasswordlessCode code)
                 sqlCon.commit();
             } catch (SQLException e) {
 
-                String message = e.getMessage();
-                if (message.contains("foreign key constraint fails")
-                        && message.contains(Config.getConfig(start).getPasswordlessCodesTable())
-                        && message.contains("device_id_hash")) {
+                if (e.getMessage().contains("foreign key constraint fails")
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java:183]: My issue was not with `message.contains(` part.. but was with your using "foreign key constraint fails" as the matching criteria instead of "foreign key" (as is done in other cases..)

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java [hunk 23]
```diff
@@ -0,0 +1,535 @@
...
+                + "link_code_salt CHAR(44) NOT NULL," + "failed_attempts INT UNSIGNED NOT NULL,"
+                + "PRIMARY KEY (device_id_hash));";
+    }
+
+    public static String getQueryToCreateCodesTable(Start start) {
+        return "CREATE TABLE IF NOT EXISTS " + Config.getConfig(start).getPasswordlessCodesTable() + " ("
+                + "code_id CHAR(36) NOT NULL," + "device_id_hash CHAR(44) NOT NULL,"
+                + "link_code_hash CHAR(44) NOT NULL UNIQUE," + "created_at BIGINT UNSIGNED NOT NULL,"
+                + "PRIMARY KEY (code_id),"
+                + "FOREIGN KEY (device_id_hash) REFERENCES passwordless_devices(device_id_hash) ON DELETE CASCADE ON UPDATE CASCADE);";
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java:59]: References can fail if user has added a prefix and has renamed `passwordless_devices` table.

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java (hunk 21 above)
Comment (siddharthmudgal) [reply src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java]: oh alright. done.

File: src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java (hunk 22 above)
Comment (siddharthmudgal) [reply src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java]: done.

File: src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java (hunk 23 above)
Comment (siddharthmudgal) [reply src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java]: done.

**Review by rishabhpoddar — APPROVED** [review]

**Review by porcellus — APPROVED** [review]
